hal/common/EventLoop.cpp

```diff
@@ -23,18 +23,23 @@
 namespace mbed {
 
 EventLoop::EventLoop(uint32_t size, osPriority priority, uint32_t stack_size, unsigned char *stack_pointer):
-    _queue(size),_thread(priority, stack_size, stack_pointer) {
+    _queue(size),_thread(priority, stack_size, stack_pointer), _stop_request(false) {
 }
 
 bool EventLoop::start() {
     return _thread.start(callback(this, &EventLoop::dispatch)) == osOK;
 }
 
-void EventLoop::stop() {
-    _thread.terminate();
+bool EventLoop::stop() {
+    _stop_request = true;
+    return _thread.join() == osOK;
 }
 
 bool EventLoop::post(const queue_event_t& evt) {
+    if (_stop_request) {
+        // An event loop that's stopping doesn't accept more events.
+        return false;
```

Comment:
One concern, this does limit resuming event queues that are acting as state machines after stopping.

``` cpp
EventQueue queue;
queue.start();                     // run for a while
wait_for_important_thing
queue.stop();                      // pause event loop for important thing that may need jitter low
do_important_thing();
queue.start();                     // resume running, unfortunately events may have been dropped during important thing
```

An alternative solution would be to poison the EventLoop with a dummy event (pardon my bad pseudocode):
``` cpp
void EventLoop::dispatch(void) {
    while (true) {
        e = _queue.get();
        if (e.value.p == EVENT_POISON) {
            break;
        }
    }
}

void EventLoop::stop(void) {
    post(EVENT_POISON);
    thread.join();
}
```